File: windows/src/Volar.App/Services/Account/AccountModels.cs
// Services/Account/AccountModels.cs — wire-adjacent DTOs and small value types for the account-auth
// feature (specs/002-workflow-command-center/contracts/account-auth.md, "CHỐT bởi anh Khôi
// 2026-07-26" — the SAME contract the macOS client and the Supabase edge functions implement in
// parallel; this file only mirrors it, never redesigns it).
//
// Layering note: every type here lives in Volar.App (not Volar.Domain) because it is consumed ONLY
// by Volar.App code (AccountService, AccountViewModel, SettingsViewModel, SettingsView.xaml.cs).
// The two lower-layer consumers that DO need a sliver of this — Volar.Parsing's
// ConfigParseCredentialProvider and Volar.Speech.Groq's AccountGroqCredentialProvider — cannot
// reference Volar.App at all (Volar.Parsing.csproj/Volar.Speech.csproj only reference
// Volar.Core/Volar.Domain or nothing; adding a reference is a csproj change outside this task's
// file-ownership scope). Those two are wired via plain `Func<...>` delegates instead (see their own
// files' doc comments) — the same "stay project-reference-free" pattern
// EnvironmentGroqCredentialProvider already established for Volar.Speech.
using System.Text.Json.Serialization;

namespace Volar.App.Services.Account;

/// <summary>Whether <see cref="AccountService"/> currently holds a usable (possibly expired-but-
/// refreshable) session. Mirrors the contract's "(chưa login)" vs account row — deliberately just
/// two states; tier (free/pro) is a SEPARATE axis carried by <see cref="SubscriptionStatusSnapshot"/>,
/// not folded into this enum, because tier can change (webhook/renewal) without ever signing out.</summary>
public enum AccountSessionState
{
    SignedOut,
    SignedIn,
}

/// <summary>The signed-in user's identity — just enough for the Settings UI to show "signed in as
/// ...". Never anything more (no PII beyond what GoTrue itself already returns).</summary>
public sealed record AccountUser(string Id, string? Email);

/// <summary>Persisted, DPAPI-encrypted session state (see <see cref="TokenStore"/>). Kept as a
/// record purely for value-equality in tests — the encrypted-at-rest guarantee comes from
/// <see cref="DpapiTokenStore"/>, not from anything about this type itself, so this type is free to
/// carry the access token in memory/JSON exactly like any other DTO.</summary>
public sealed record StoredSession(
    string AccessToken,
    string RefreshToken,
    DateTimeOffset ExpiresAt,
    string UserId,
    string? Email);

/// <summary>Cached snapshot of <c>GET /functions/v1/subscription/status</c> (contract §3) plus the
/// instant it was fetched, so callers (the Groq pre-record gate, the Settings quota line) can judge
/// staleness themselves rather than this type silently going stale in the background.</summary>
public sealed record SubscriptionStatusSnapshot(
    string Tier,
    DateTimeOffset? ExpiresAt,
    int ParseUsedToday,
    int ParseLimit,
    int SpeechUsedToday,
    int SpeechLimit,
    DateTimeOffset FetchedAt)
{
    public bool IsPro => Tier == "pro";
}

// MARK: - Outcome types for the auth-flow entry points (VM-facing, richer than a plain bool so the
// UI can show a useful inline message per the contract's opaque {"error":"<code>"} error shape).

public enum OtpSendOutcome
{
    Sent,
    InvalidEmail,
    NetworkError,
    ServerError,
}

public sealed record OtpSendResult(OtpSendOutcome Outcome, string? ErrorMessage = null)
{
    public static OtpSendResult Ok() => new(OtpSendOutcome.Sent);
}

public enum VerifyOtpOutcome
{
    SignedIn,
    InvalidCode,
    NetworkError,
    ServerError,
}

public sealed record VerifyOtpResult(VerifyOtpOutcome Outcome, string? ErrorMessage = null);

public enum DeleteAccountOutcome
{
    Deleted,
    NetworkError,
    ServerError,
    NotSignedIn,
}

public sealed record DeleteAccountResult(DeleteAccountOutcome Outcome, string? ErrorMessage = null);

// MARK: - Wire DTOs (System.Text.Json + [JsonPropertyName], same style CloudParser/
// GroqTranscriptionClient already use for their own response types).

/// <summary>Supabase GoTrue's user object, as embedded in a session response.</summary>
internal sealed class GoTrueUserWire
{
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("email")]
    public string? Email { get; set; }
}

/// <summary>Supabase GoTrue's session response — contract §2: "Session trả về: {access_token,
/// refresh_token, expires_at (epoch giây), token_type, user:{id, email}}". Used for both the
/// verify-OTP response and the refresh-token response (identical shape).</summary>
internal sealed class GoTrueSessionWire
{
    [JsonPropertyName("access_token")]
    public string? AccessToken { get; set; }

    [JsonPropertyName("refresh_token")]
    public string? RefreshToken { get; set; }

    /// <summary>Epoch SECONDS per the contract ("expires_at (epoch giây)") — NOT milliseconds.</summary>
    [JsonPropertyName("expires_at")]
    public long? ExpiresAt { get; set; }

    [JsonPropertyName("user")]
    public GoTrueUserWire? User { get; set; }
}

/// <summary>Wire shape of <c>GET /functions/v1/subscription/status</c> (contract §3) —
/// genuinely camelCase on the wire, unlike GoTrue's own snake_case.</summary>
internal sealed class SubscriptionStatusWire
{
    [JsonPropertyName("tier")]
    public string? Tier { get; set; }

    [JsonPropertyName("expiresAt")]
    public string? ExpiresAt { get; set; }

    [JsonPropertyName("parseUsedToday")]
    public int ParseUsedToday { get; set; }

    [JsonPropertyName("parseLimit")]
    public int ParseLimit { get; set; }

    [JsonPropertyName("speechUsedToday")]
    public int SpeechUsedToday { get; set; }

    [JsonPropertyName("speechLimit")]
    public int SpeechLimit { get; set; }
}

File: Shared/Account/AccountModels.swift
// Sources/Account/AccountModels.swift — wire types for Supabase Auth + entitlements/quota.
// specs/002-workflow-command-center/contracts/account-auth.md §2 (auth), §3 (edge functions), §6.
//
// Plain Codable/Sendable value types only — no networking, no Keychain, no StoreKit here (see
// `AccountService.swift`/`Entitlements.swift`/`KeychainStore.swift` for those). Kept in one file
// so every collaborator shares exactly one decode shape per wire response.
import Foundation

/// A Supabase Auth session — the exact shape every GoTrue endpoint this app calls returns
/// (`/auth/v1/token`, `/auth/v1/verify`), per contract §2: "Session trả về: {access_token,
/// refresh_token, expires_at (epoch giây), token_type, user:{id, email}}".
///
/// This is BOTH the wire-decode shape AND the Keychain-persisted shape (same `Codable`, JSON-
/// encoded as one blob) — see `KeychainStore.swift`'s doc comment for why one item is enough.
struct AccountSession: Codable, Sendable, Equatable {
    var accessToken: String
    var refreshToken: String
    /// Epoch SECONDS (contract: "expires_at (epoch giây)") — GoTrue sends a raw JSON number, not
    /// an ISO8601 string, so this decodes as `Double` rather than through a date strategy that
    /// would apply to every field in this struct. Use `expiresAtDate` below to work with `Date`.
    var expiresAt: Double
    var tokenType: String
    var user: AccountUser

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresAt = "expires_at"
        case tokenType = "token_type"
        case user
    }

    var expiresAtDate: Date { Date(timeIntervalSince1970: expiresAt) }
}

struct AccountUser: Codable, Sendable, Equatable {
    var id: String
    /// Contract §2: "Apple chỉ trả email lần đăng nhập ĐẦU TIÊN" — Apple's identity token only
    /// carries an email claim on first authorization, but Supabase persists it into `auth.users`
    /// server-side, so every session response (first login or not) is expected to still carry it
    /// here. Modeled optional anyway to be defensive against a GoTrue response that omits it
    /// (e.g. an email-less account edge case) rather than force-unwrapping.
    var email: String?
}

/// `GET /functions/v1/subscription/status` 200 body — contract §3.
struct SubscriptionStatus: Codable, Sendable, Equatable {
    var tier: AccountTier
    var expiresAt: String?
    var parseUsedToday: Int
    var parseLimit: Int
    var speechUsedToday: Int
    var speechLimit: Int
}

/// `POST /functions/v1/subscription/link` 200 body — contract §3.
struct SubscriptionLinkResponse: Codable, Sendable, Equatable {
    var tier: AccountTier
    var expiresAt: String?
    var productId: String?
}

enum AccountTier: String, Codable, Sendable, Equatable {
    case free, pro
}

/// Opaque server error body, contract §5: `{"error":"<code>"}` (+ `resetAt` on 429, unused by
/// this subsystem — quota 429s are `CloudParser`'s/`GroqTranscriptionClient`'s concern, not
/// account auth's).
struct AccountErrorBody: Codable, Sendable {
    var error: String
    var resetAt: String?
}

/// Every failure `AccountService`/`Entitlements` can produce — ONE enum (not split per file) so
/// `SettingsView` has a single error type to switch/display. Mirrors the error-enum convention of
/// `GroqTranscriptionError`/`ParseCredentialError` elsewhere in this codebase.
enum AccountError: Error, Sendable, Equatable {
    /// No session, or a refresh attempt failed — the "please sign in" state. IMPORTANT: this must
    /// NEVER reach `ConfigParseCredentialProvider`/`GroqTranscriptionClient`'s credential
    /// providers as a surfaced error — they treat "no valid token" as "return nil credential" and
    /// fall back on-device silently (contract §1: "Chưa login → ... Không hiện lỗi"). Only
    /// `SettingsView`'s explicit sign-in flow shows this to the user.
    case signedOut
    /// User dismissed/cancelled the Apple Sign-In sheet — not a real failure.
    case cancelled
    /// Non-2xx HTTP response. `code` is the decoded `{"error":"…"}` body when present (contract
    /// §5's opaque error codes), `nil` if the body didn't decode.
    case http(status: Int, code: String?)
    case decoding
    case network(String)
}

extension AccountError: LocalizedError {
    var errorDescription: String? {
        switch self {
        case .signedOut: return "Not signed in."
        case .cancelled: return "Sign-in was cancelled."
        case .http(let status, let code): return "Request failed (\(code ?? "http_\(status)"))."
        case .decoding: return "Couldn't read the server's response."
        case .network(let message): return "Network error: \(message)"
        }
    }
}
